flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java

```diff
@@ -225,13 +243,78 @@ public static JobGraph createNonEmptyJobGraph() {
 		return jobGraph;
 	}
 
-	private static class HATestingDispatcher extends TestingDispatcher {
+	@Test
+	public void testPersistedJobGraphWhenDispatcherIsShutDown() throws Exception {
+		final JobGraph jobGraph = createNonEmptyJobGraph();
+		final TestingHighAvailabilityServices highAvailabilityServices = new TestingHighAvailabilityServices();
 
-		@Nonnull
-		private final BlockingQueue<DispatcherId> fencingTokens;
+		final InMemorySubmittedJobGraphStore submittedJobGraphStore = new InMemorySubmittedJobGraphStore();
+		highAvailabilityServices.setSubmittedJobGraphStore(submittedJobGraphStore);
+
+		final TestingLeaderElectionService leaderElectionService = new TestingLeaderElectionService();
+		highAvailabilityServices.setDispatcherLeaderElectionService(leaderElectionService);
+
+		highAvailabilityServices.setJobMasterLeaderElectionService(jobGraph.getJobID(), new TestingLeaderElectionService());
+		highAvailabilityServices.setResourceManagerLeaderRetriever(new SettableLeaderRetrievalService());
+		highAvailabilityServices.setCheckpointRecoveryFactory(new StandaloneCheckpointRecoveryFactory());
 
-		HATestingDispatcher(RpcService rpcService, String endpointId, Configuration configuration, HighAvailabilityServices highAvailabilityServices, ResourceManagerGateway resourceManagerGateway, BlobServer blobServer, HeartbeatServices heartbeatServices, JobManagerMetricGroup jobManagerMetricGroup, @Nullable String metricQueryServicePath, ArchivedExecutionGraphStore archivedExecutionGraphStore, JobManagerRunnerFactory jobManagerRunnerFactory, FatalErrorHandler fatalErrorHandler, @Nonnull BlockingQueue<DispatcherId> fencingTokens) throws Exception {
-			super(rpcService, endpointId, configuration, highAvailabilityServices, resourceManagerGateway, blobServer, heartbeatServices, jobManagerMetricGroup, metricQueryServicePath, archivedExecutionGraphStore, jobManagerRunnerFactory, fatalErrorHandler);
+		final HATestingDispatcher dispatcher = createDispatcher(
```

Comment:
This test does not seem to need `fencingTokens`. 
At the first glance, does `getNumberJobs` has any HA specifics? Could it also belong to `TestingDispatcher`?
If it is true then this test could just use `TestingDispatcher`, although, constructing code would need to be duplicated.

A bit off topic, do we need the `VisibleForTesting` annotation for this method, as it already resides in test code?

Reply:
I think you're right and the test should be moved to `DispatcherTest` because it is not HA specific. Moreover, we should move `getNumberJobs` into `TestingDispatcher`.

The annotation `VisibleForTesting` is not necessary since `HATestingDispatcher` is a testing class. Will remove it while moving the method over to `TestingDispatcher`.